Documentation/man/man3/AWEGlobalSettings.3
.TH "AWEGlobalSettings" 3 "Thu Apr 17 2014" "Version 0.1" "AWE Media Center" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AWEGlobalSettings \- 
.PP
Holds the global settings for AWEMC\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <AWEGlobalSettings\&.h>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "typedef std::set< std::string > \fBNameSet\fP"
.br
.RI "\fIThe set type used to hold names\&. \fP"
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAWEGlobalSettings\fP (const std::string &settingsFile)"
.br
.RI "\fICreate the global settings object\&. \fP"
.ti -1c
.RI "\fB~AWEGlobalSettings\fP ()"
.br
.RI "\fIDeconstructor\&. \fP"
.ti -1c
.RI "\fBAWEMetadataScraper\fP * \fBgetScraperByName\fP (const std::string &name)"
.br
.RI "\fIGet a metadata scraper by name\&. \fP"
.ti -1c
.RI "\fBJson::Value\fP & \fBgetScraperSettingsByName\fP (const std::string &name)"
.br
.RI "\fIGet a metadata scraper's settings by name\&. \fP"
.ti -1c
.RI "const \fBNameSet\fP & \fBgetAllMetadataScraperNames\fP ()"
.br
.RI "\fIGet a set of all metadata scrapers\&. \fP"
.ti -1c
.RI "\fBAWEMediaPlayer\fP * \fBgetPlayerByName\fP (const std::string &name)"
.br
.RI "\fIGet a media player by name\&. \fP"
.ti -1c
.RI "\fBJson::Value\fP & \fBgetPlayerSettingsByName\fP (const std::string &name)"
.br
.RI "\fIGet a media player's settings by name\&. \fP"
.ti -1c
.RI "const \fBNameSet\fP & \fBgetAllMediaPlayerNames\fP ()"
.br
.RI "\fIGet a set of all media players\&. \fP"
.ti -1c
.RI "\fBJson::Value\fP & \fBgetTypeByName\fP (const std::string &name)"
.br
.RI "\fIGet the media type with the given name\&. \fP"
.ti -1c
.RI "const \fBNameSet\fP & \fBgetAllMediaTypeNames\fP ()"
.br
.RI "\fIGet a set of all media types\&. \fP"
.ti -1c
.RI "const \fBNameSet\fP & \fBgetAllMediaServiceNames\fP ()"
.br
.RI "\fIGet the names of all of the media services\&. \fP"
.ti -1c
.RI "\fBAWEMediaService\fP * \fBgetMediaServiceByName\fP (const std::string &name)"
.br
.RI "\fIGet a media service by name\&. \fP"
.ti -1c
.RI "\fBAWEMediaItem\fP * \fBgetMediaItemByJSONFile\fP (const QDir &file)"
.br
.RI "\fIGet a media item from its JSON file\&. \fP"
.ti -1c
.RI "void \fBaddFolder\fP (const std::string &path, \fBAWEFolder\fP *folder)"
.br
.RI "\fIAdd the given folder\&. \fP"
.ti -1c
.RI "\fBAWEFolder\fP * \fBgetRootFolder\fP ()"
.br
.RI "\fIGet the root folder\&. \fP"
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static \fBJson::Value\fP \fBnull\fP = Json::Value::null"
.br
.in -1c
.SH "Detailed Description"
.PP 
Holds the global settings for AWEMC\&. 


.PP
\fBTodo\fP
.RS 4
Skins\&.\&.\&. 
.PP
Get a better folder adding system\&.
.RE
.PP
.PP
Functions for pretty much anything you want to know about the prefernces for AWEMC are here\&. Every plugin extending AWEMC has its own configuration file, which is held in here via sub-configuration objects\&.
.PP
You shouldn't need to instantiate this class on your own; most objects are passed a copy of the settings object during creation or through some other means\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AWEGlobalSettings::AWEGlobalSettings (const std::string &settingsFile)"

.PP
Create the global settings object\&. 
.PP
\fBParameters:\fP
.RS 4
\fIsettingsFile\fP The file that contains the settings for AWEMC (usually \fCsettings\&.json\fP in the root folder) 
.RE
.PP

.SS "AWEGlobalSettings::~AWEGlobalSettings ()"

.PP
Deconstructor\&. Deletes every scraper, player and item\&. 
.SH "Member Function Documentation"
.PP 
.SS "void AWEGlobalSettings::addFolder (const std::string &path, \fBAWEFolder\fP *folder)"

.PP
Add the given folder\&. 
.PP
\fBParameters:\fP
.RS 4
\fIpath\fP The path to the config file\&. 
.br
\fIfolder\fP The folder\&. 
.RE
.PP

.SS "const \fBAWEGlobalSettings::NameSet\fP & AWEGlobalSettings::getAllMediaPlayerNames ()"

.PP
Get a set of all media players\&. 
.PP
\fBReturns:\fP
.RS 4
A set filled with every media player name\&. 
.RE
.PP

.SS "const \fBAWEGlobalSettings::NameSet\fP & AWEGlobalSettings::getAllMediaServiceNames ()"

.PP
Get the names of all of the media services\&. 
.PP
\fBReturns:\fP
.RS 4
A set filled with every media service name\&. 
.RE
.PP

.SS "const \fBAWEGlobalSettings::NameSet\fP & AWEGlobalSettings::getAllMediaTypeNames ()"

.PP
Get a set of all media types\&. 
.PP
\fBReturns:\fP
.RS 4
A set filled with every media type name\&. 
.RE
.PP

.SS "const \fBAWEGlobalSettings::NameSet\fP & AWEGlobalSettings::getAllMetadataScraperNames ()"

.PP
Get a set of all metadata scrapers\&. 
.PP
\fBReturns:\fP
.RS 4
A set filled with every metadata scraper name\&. 
.RE
.PP

.SS "\fBAWEMediaItem\fP * AWEGlobalSettings::getMediaItemByJSONFile (const QDir &file)"

.PP
Get a media item from its JSON file\&. \fCfile\fP is a \fCQDir\fP, so relative paths and links do not duplicate\&. If \fCfile\fP is not found, it is added\&.
.PP
\fBReturns:\fP
.RS 4
The desired media item\&. 
.RE
.PP

.SS "\fBAWEMediaService\fP * AWEGlobalSettings::getMediaServiceByName (const std::string &name)"

.PP
Get a media service by name\&. 
.PP
\fBParameters:\fP
.RS 4
\fIname\fP The name of the media service\&.
.RE
.PP
\fBReturns:\fP
.RS 4
The desired media service\&. 
.RE
.PP

.SS "\fBAWEMediaPlayer\fP * AWEGlobalSettings::getPlayerByName (const std::string &name)"

.PP
Get a media player by name\&. 
.PP
\fBParameters:\fP
.RS 4
\fIname\fP The name of the player\&.
.RE
.PP
\fBReturns:\fP
.RS 4
The desired media player or \fCNULL\fP if it does not exist\&. 
.RE
.PP

.SS "\fBJson::Value\fP & AWEGlobalSettings::getPlayerSettingsByName (const std::string &name)"

.PP
Get a media player's settings by name\&. 
.PP
\fBParameters:\fP
.RS 4
\fIname\fP The name of the player\&.
.RE
.PP
\fBReturns:\fP
.RS 4
The settings of the desired media player or \fCJson::Value::null\fP if it does not exist\&. 
.RE
.PP

.SS "\fBAWEFolder\fP * AWEGlobalSettings::getRootFolder ()"

.PP
Get the root folder\&. 
.PP
\fBReturns:\fP
.RS 4
The root folder\&. 
.RE
.PP

.SS "\fBAWEMetadataScraper\fP * AWEGlobalSettings::getScraperByName (const std::string &name)"

.PP
Get a metadata scraper by name\&. 
.PP
\fBParameters:\fP
.RS 4
\fIname\fP The name of the scraper\&.
.RE
.PP
\fBReturns:\fP
.RS 4
The desired scraper as an \fC\fBAWEMetadataScraper\fP\fP object or \fCNULL\fP if it does not exist\&. 
.RE
.PP

.SS "\fBJson::Value\fP & AWEGlobalSettings::getScraperSettingsByName (const std::string &name)"

.PP
Get a metadata scraper's settings by name\&. 
.PP
\fBParameters:\fP
.RS 4
\fIname\fP The name of the scraper\&.
.RE
.PP
\fBReturns:\fP
.RS 4
The settings of the desired scraper or \fCJson::Value::null\fP if it does not exist\&. 
.RE
.PP

.SS "\fBJson::Value\fP & AWEGlobalSettings::getTypeByName (const std::string &name)"

.PP
Get the media type with the given name\&. 
.PP
\fBReturns:\fP
.RS 4
The default metadata settings for the given type\&. 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for AWE Media Center from the source code\&.


Code/settings/AWEGlobalSettings.cpp
// class header
#include "AWEGlobalSettings.h"

// the types that are held inside GlobalSettings

// scrapers
#include "scraper/AWEMetadataScraper.h"
#include "scraper/AWEJSONScraper.h"

// players
#include "player/AWEMediaPlayer.h"
#include "player/AWEJSONPlayer.h"

// items
#include "items/AWEMediaItem.h"
#include "items/AWEMediaFile.h"
#include "items/AWEFolder.h"
#include "items/AWEMediaService.h"
#include "items/AWEJSONService.h"

// for traversing paths
#include <QDir>
#include <QStringList>
#include <QString>

// for reading files
#include "libs/generic_file_reader/file_reader.h"
#include <iostream>
#include <sstream>

using namespace AWE;
using namespace Json;
using namespace std;

Value GlobalSettings::null = Value::null;

GlobalSettings::GlobalSettings(const string& settingsFile)
{
	// TODO make safe
	stringstream ss;
	copyFile(settingsFile, ss);
	Reader reader;
	reader.parse(ss.str(), mySettingsFile, false);

	// get all of the objects
	obtainScrapers();
	obtainPlayers();
	obtainSkins();
	obtainTypes();
	obtainServices();
	obtainItems();
}

GlobalSettings::~GlobalSettings()
{
	for (auto i : myMediaPlayers)
	{
		delete i.second;
	}
	for (auto i : myMetadataScrapers)
	{
		delete i.second;
	}
	for (auto i : myMediaItems)
	{
		delete i.second;
	}
}

void GlobalSettings::obtainScrapers()
{
	// get the directory
	QDir dir(QDir::current());
	dir.cd(mySettingsFile["scrapers"].asCString());

	// important variables
	Reader reader;
	Value scraper;
	string name, type, file;
	stringstream contents;

	// get all of the JSON scrapers
	dir.cd("json");
	QStringList filter;
	filter << "*.json";
	dir.setNameFilters(filter);
	for (unsigned int i = 0; i < dir.count(); ++ i)
	{
		file = dir.absoluteFilePath(dir[i]).toStdString();
		copyFile(file, contents);
		if (reader.parse(contents.str(), 
							scraper, false))
		{
			name = scraper["name"].asString();
			type = scraper["type"].asString();
			myMetadataScraperNames.insert(name);
			myMetadataScraperSettings[name] = scraper;
			myMetadataScrapers[name] = new JSONScraper(name, type);
		}
		contents.clear();
	}

	// TODO plugins
}

void GlobalSettings::obtainPlayers()
{
	// get the directory
	QDir dir(QDir::current());
	dir.cd(mySettingsFile["players"].asCString());

	// important variables
	Reader reader;
	Value player;
	string name, file;
	stringstream contents;

	// get all of the JSON players
	dir.cd("json");
	QStringList filter;
	filter << "*.json";
	dir.setNameFilters(filter);
	for (unsigned int i = 0; i < dir.count(); ++ i)
	{
		file = dir.absoluteFilePath(dir[i]).toStdString();
		copyFile(file, contents);
		if (reader.parse(contents.str(),
							player, false))
		{
			name = player["name"].asString();
			myMediaPlayerNames.insert(name);
			myMediaPlayerSettings[name] = player;
			myMediaPlayers[name] = new JSONPlayer(player);
		}
		contents.clear();
	}

	// TODO plugins
}

void GlobalSettings::obtainSkins() {}

void GlobalSettings::obtainTypes()
{
	QDir dir(QDir::current());
	dir.cd(mySettingsFile["types"].asCString());

	// important variables
	Reader reader;
	Value type;
	string name, file;
	stringstream contents;

	// get all of the media types
	QStringList filter;
	filter << "*.json";
	dir.setNameFilters(filter);
	for (unsigned int i = 0; i < dir.count(); ++ i)
	{
		file = dir.absoluteFilePath(dir[i]).toStdString();
		copyFile(file, contents);
		if (reader.parse(contents.str(),
							type, false))
		{
			name = type["type"].asString();
			myMediaTypeNames.insert(name);
			myMediaTypes[name] = type;
		}
		contents.clear();
	}
}

void GlobalSettings::obtainServices()
{
	// get the directory
	QDir dir(QDir::current());
	dir.cd(mySettingsFile["services"].asCString());

	// important variables
	Reader reader;
	Value service;
	string name, file;
	stringstream contents;

	// get all of the JSON players
	dir.cd("json");
	QStringList filter;
	filter << "*.json";
	dir.setNameFilters(filter);
	for (unsigned int i = 0; i < dir.count(); ++ i)
	{
		file = dir.absoluteFilePath(dir[i]).toStdString();
		copyFile(file, contents);
		if (reader.parse(contents.str(),
							service, false))
		{
			name = service["name"].asString();
			myMediaServiceNames.insert(name);
			myMediaServices[name] = new JSONService(QDir(file.c_str()));
		}
		contents.clear();
	}

	// TODO plugins
}

void GlobalSettings::obtainItems() 
{
	QDir root(mySettingsFile["root"].asCString());
	myRootFolder = (Folder*) getMediaItemByJSONFile(root);
}

MetadataScraper* GlobalSettings::getScraperByName(const string& str)
{
	map<string, MetadataScraper*>::iterator i
		= myMetadataScrapers.find(str);
	if (i != myMetadataScrapers.end())
	{
		return i->second;
	}
	return NULL;
}

Value& GlobalSettings::getScraperSettingsByName(const string& str)
{
	map<string, Value>::iterator i
		= myMetadataScraperSettings.find(str);
	if (i != myMetadataScraperSettings.end())
	{
		return i->second;
	}
	return null;
}

const GlobalSettings::NameSet& GlobalSettings::getAllMetadataScraperNames()
{
	return myMetadataScraperNames;
}

MediaPlayer* GlobalSettings::getPlayerByName(const string& str)
{
	map<string, MediaPlayer*>::iterator i
		= myMediaPlayers.find(str);
	if (i != myMediaPlayers.end())
	{
		return i->second;
	}
	return NULL;
}

Value& GlobalSettings::getPlayerSettingsByName(const string& str)
{
	map<string, Value>::iterator i
		= myMediaPlayerSettings.find(str);
	if (i != myMediaPlayerSettings.end())
	{
		return i->second;
	}
	return null;
}

const GlobalSettings::NameSet& GlobalSettings::getAllMediaPlayerNames()
{
	return myMediaPlayerNames;
}

Value& GlobalSettings::getTypeByName(const string& str)
{
	map<string, Value>::iterator i
		= myMediaTypes.find(str);
	if (i != myMediaTypes.end())
	{
		return i->second;
	}
	return null;
}

const GlobalSettings::NameSet& GlobalSettings::getAllMediaTypeNames()
{
	return myMediaTypeNames;
}

MediaItem* GlobalSettings::getMediaItemByJSONFile(const QDir& file)
{
	string str = QDir::cleanPath(file.absolutePath()).toStdString();
	map<string, MediaItem*>::iterator i = myMediaItems.find(str);
	if (i != myMediaItems.end())
	{
		return i->second;
	}
	else
	{
		// read to get the type
		MediaItem* item = NULL;
		Reader reader;
		Value temp;
		stringstream ss;
		if (copyFile(file.absolutePath().toStdString(), ss)
			&& reader.parse(ss.str(), temp, false))
		{
			if (temp["type"].asString() == "file")
			{
				myMediaItems[str] = (item = new MediaFile(file, this));
			}
			else if (temp["type"].asString() == "folder")
			{
				item = new Folder(file, this);
			}
			else if (temp["type"].asString() == "service")
			{
				item = getMediaServiceByName(temp["name"].asString());
			}
		}
		return item;
	}
}

const GlobalSettings::NameSet& GlobalSettings::getAllMediaServiceNames()
{
	return myMediaServiceNames;
}

MediaService* GlobalSettings::getMediaServiceByName(const string& str)
{
	map<string, MediaService*>::iterator i
		= myMediaServices.find(str);
	if (i != myMediaServices.end())
	{
		return i->second;
	}
	return NULL;
}

void GlobalSettings::addFolder(const string& str, Folder* folder)
{
	myMediaItems[str] = folder;
}

Folder* GlobalSettings::getRootFolder()
{
	return myRootFolder;
}